packages/api/shared/access-control/authorize.ts

```diff
@@ -0,0 +1,77 @@
+import { subject as buildSubject } from '@casl/ability';
+import type { RouteRequest } from '../Route';
+import { client } from '../db';
+import { ForbiddenError, NotFoundError } from '../errors';
+import { Action, createPolicyFor, RawSubject, Subject } from './policy';
+
+interface AuthorizeOptions {
+  action: Action;
+  subject: RawSubject;
+  subjectId?: string;
+}
+
+/**
+ * Determine whether a user is authorized to perform the requested action on the given subject.
+ *
+ * If no subject ID is given, the user will be authorized if they have access to _any_ (not all) of that subject type.
+ * @param options
+ */
+export function authorize<Params, Body>(
+  options:
+    | AuthorizeOptions
+    | ((req: RouteRequest<Params, Body>) => AuthorizeOptions)
+): (req: RouteRequest<Params, Body>) => Promise<void> {
+  return async (req: RouteRequest<Params, Body>) => {
+    const config = typeof options === 'object' ? options : options(req);
+
+    const policy = createPolicyFor(req.session?.user);
+
+    let subject: Subject | undefined = undefined;
+    if ('subjectId' in config && config.subjectId) {
+      switch (config.subject) {
+        case 'Language': {
+          const language = await client.language.findUnique({
+            where: { code: config.subjectId },
+          });
```

Comment:
In the future we will want to load language members so that the policy can test against that data as well